File: java/BillsApi/src/main/java/com/apilangarena/javabillsapi/presentation/ApiExceptionHandler.java
package com.apilangarena.javabillsapi.presentation;

import com.apilangarena.javabillsapi.application.common.ConflictException;
import com.apilangarena.javabillsapi.domain.common.DomainValidationException;
import org.springframework.amqp.AmqpException;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.http.converter.HttpMessageNotReadableException;
import org.springframework.validation.FieldError;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

@RestControllerAdvice
public class ApiExceptionHandler {
    @ExceptionHandler(MethodArgumentNotValidException.class)
    public ResponseEntity<ProblemResponse> handleMethodArgumentNotValid(MethodArgumentNotValidException ex) {
        Map<String, List<String>> fieldErrors = new LinkedHashMap<>();
        for (var error : ex.getBindingResult().getAllErrors()) {
            String fieldName = error instanceof FieldError fe ? fe.getField() : error.getObjectName();
            String message = error.getDefaultMessage() == null ? "Invalid value" : error.getDefaultMessage();
            fieldErrors.computeIfAbsent(fieldName, ignored -> new ArrayList<>()).add(message);
        }

        return problem(HttpStatus.BAD_REQUEST, "Validation failed", fieldErrors);
    }

    @ExceptionHandler(HttpMessageNotReadableException.class)
    public ResponseEntity<ProblemResponse> handleInvalidJson(HttpMessageNotReadableException ignored) {
        return problem(HttpStatus.BAD_REQUEST, "Validation failed", Map.of("request", List.of("Invalid JSON payload.")));
    }

    @ExceptionHandler(DomainValidationException.class)
    public ResponseEntity<ProblemResponse> handleDomainValidation(DomainValidationException ex) {
        return problem(HttpStatus.BAD_REQUEST, "Validation failed", ex.getErrors());
    }

    @ExceptionHandler(ConflictException.class)
    public ResponseEntity<ProblemResponse> handleConflict(ConflictException ex) {
        return problem(HttpStatus.CONFLICT, ex.getMessage(), null);
    }

    @ExceptionHandler(AmqpException.class)
    public ResponseEntity<ProblemResponse> handleAmqp(AmqpException ex) {
        return problem(HttpStatus.SERVICE_UNAVAILABLE, "Message broker error: " + ex.getClass().getSimpleName(), null);
    }

    @ExceptionHandler(Exception.class)
    public ResponseEntity<ProblemResponse> handleUnhandled(Exception ignored) {
        return problem(HttpStatus.INTERNAL_SERVER_ERROR, "An unexpected error occurred.", null);
    }

    private ResponseEntity<ProblemResponse> problem(HttpStatus status, String title, Map<String, List<String>> errors) {
        return ResponseEntity
            .status(status)
            .body(new ProblemResponse("about:blank", title, status.value(), errors));
    }
}

File: kotlin/BillsApi/src/main/kotlin/com/apilangarena/billsapi/presentation/ApiExceptionHandler.kt
package com.apilangarena.billsapi.presentation

import com.apilangarena.billsapi.application.common.ConflictException
import com.apilangarena.billsapi.domain.common.DomainValidationException
import org.springframework.amqp.AmqpException
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.http.converter.HttpMessageNotReadableException
import org.springframework.validation.FieldError
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice

data class ProblemResponse(
    val type: String = "about:blank",
    val title: String,
    val status: Int,
    val errors: Map<String, List<String>>? = null,
)

@RestControllerAdvice
class ApiExceptionHandler {
    @ExceptionHandler(MethodArgumentNotValidException::class)
    fun handleMethodArgumentNotValid(ex: MethodArgumentNotValidException): ResponseEntity<ProblemResponse> {
        val fieldErrors = ex.bindingResult
            .allErrors
            .groupBy { error ->
                when (error) {
                    is FieldError -> error.field
                    else -> error.objectName
                }
            }
            .mapValues { (_, errors) ->
                errors.mapNotNull { it.defaultMessage }.ifEmpty { listOf("Invalid value") }
            }

        return problem(HttpStatus.BAD_REQUEST, "Validation failed", fieldErrors)
    }

    @ExceptionHandler(HttpMessageNotReadableException::class)
    fun handleInvalidJson(ex: HttpMessageNotReadableException): ResponseEntity<ProblemResponse> {
        return problem(
            HttpStatus.BAD_REQUEST,
            "Validation failed",
            mapOf("request" to listOf("Invalid JSON payload.")),
        )
    }

    @ExceptionHandler(DomainValidationException::class)
    fun handleDomainValidation(ex: DomainValidationException): ResponseEntity<ProblemResponse> {
        return problem(HttpStatus.BAD_REQUEST, "Validation failed", ex.errors)
    }

    @ExceptionHandler(ConflictException::class)
    fun handleConflict(ex: ConflictException): ResponseEntity<ProblemResponse> {
        return problem(HttpStatus.CONFLICT, ex.message ?: "Conflict", null)
    }

    @ExceptionHandler(AmqpException::class)
    fun handleAmqp(ex: AmqpException): ResponseEntity<ProblemResponse> {
        return problem(
            HttpStatus.SERVICE_UNAVAILABLE,
            "Message broker error: ${ex.javaClass.simpleName}",
            null,
        )
    }

    @ExceptionHandler(Exception::class)
    fun handleUnhandled(ex: Exception): ResponseEntity<ProblemResponse> {
        return problem(HttpStatus.INTERNAL_SERVER_ERROR, "An unexpected error occurred.", null)
    }

    private fun problem(
        status: HttpStatus,
        title: String,
        errors: Map<String, List<String>>?,
    ): ResponseEntity<ProblemResponse> {
        return ResponseEntity
            .status(status)
            .body(
                ProblemResponse(
                    title = title,
                    status = status.value(),
                    errors = errors,
                )
            )
    }
}
